lib/src/puppetdb-cli.cc

```diff
@@ -107,7 +121,84 @@ PuppetDBConn::parseServerUrls(const json::JsonContainer& config) {
     }
 }
 
-size_t write_data(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+class SynchronizedCharQueue {
+  public:
+    vector<char> front() {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        cond_.wait(mlock,  [&]{ return !queue_.empty(); });
+        return queue_.front();
+    }
+
+    vector<char> pop() {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        cond_.wait(mlock,  [&]{ return !queue_.empty(); });
+        auto item = queue_.front();
+        queue_.pop();
+        return item;
+    }
+
+    void push(vector<char> items) {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        queue_.push(items);
+        mlock.unlock();
+        cond_.notify_one();
+    }
+
+    void push(char item) {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        queue_.push({item});
+        mlock.unlock();
+        cond_.notify_one();
+    }
+
+  private:
+    queue< vector<char> > queue_;
+    boost::mutex mutex_;
+    boost::condition_variable cond_;
+};
+
+struct UserData {
+    bool collecting_header;
+    vector<string> headers;
+    string content_type;
+    string error_response;
+    SynchronizedCharQueue stream;
+};
+
+size_t write_queue(char *ptr, size_t size, size_t nmemb, UserData& userdata) {
+    const size_t written = size * nmemb;
+    if (userdata.collecting_header) {
+        if (written == 2 && ptr[0] == '\r' && ptr[1] == '\n') {
+            // We've reached the end of the header
+            userdata.collecting_header = false;
+            return written;
+        } else {
+            boost::string_ref header(ptr, written);
+            userdata.headers.push_back(header.to_string());
+
+            if (!header.starts_with("HTTP/")) {
+                auto pos = header.find_first_of(':');
+                if (pos == boost::string_ref::npos) {
+                    LOG_WARNING("unexpected HTTP response header: {1}.", header);
+                } else {
+                    auto name = header.substr(0, pos).to_string();
+                    boost::trim(name);
+                    if (name == "Content-Type") {
+                        auto value = header.substr(pos + 1).to_string();
+                        boost::trim(value);
+                        userdata.content_type = value;
+                    }
+                }
+            }
+        }
+    } else {
+        vector<char> buffer(ptr, ptr + written);
+        userdata.stream.push(buffer);
```

Comment:
I don't think we wan't a circular_buffer, we would then have to block writes when the buffer is full and we'd have to figure out how big the buffer should be. We have no garuntees from libcurl about how much data we will be writing here. We can avoid the copy without using iterators, we just use `const vector<char>&` in the push function.